File: 1122Programming/leetcode/234. Palindrome Linked List/ans.rs
// https://leetcode.com/problems/palindrome-linked-list/
// Problem ID: 234. Palindrome Linked List
//
// Author: revival0728
//
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        let mut ptr = head;
        let mut stk = Vec::new();
        
        while let Some(cur) = ptr {
            stk.push(cur.val);
            ptr = cur.next;
        }
        for (i, r) in stk.iter().zip(stk.iter().rev()) {
            if i != r {
                return false
            }
        }
        true
    }
}

File: 1122Programming/leetcode/234. Palindrome Linked List/ans.cpp
/*
 *
 * https://leetcode.com/problems/palindrome-linked-list
 * Problem ID: 234. Palindrome Linked List
 * 
 * Author: Ateto
 * 
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        vector<int> vec;
        ListNode* dummy = head;
        while(dummy != NULL) {
            vec.push_back(dummy->val);
            dummy = dummy->next;
        }
        int n = vec.size();
        for(int i = 0; i < n; i++) {
            if(vec[i] != vec[n - i - 1]) {
                return false;
            }
        }
        return true;
    }
};